src/main/java/net/neoforged/neoforge/transfer/ResourceHandlerUtil.java

```diff
@@ -0,0 +1,565 @@
+            int amount,
+            R defaultResource,
+            @Nullable TransactionContext transaction,
+            IStackFactory<R, S> stackFactory) {
+        Objects.requireNonNull(filter, "Filter may not be null");
+
+        if (from == null || to == null)
+            return stackFactory.create(defaultResource, 0);
+
+        try {
+            int totalMoved = 0;
+            R lastMovedResource = defaultResource;
+
+            int size = from.size();
+
+            for (int index = 0; index < size; ++index) {
+                R fromResource = from.getResource(index);
+                if (fromResource.isEmpty() || !filter.test(fromResource)) continue;
+
+                // check how much can be extracted
+                int extracted;
+                try (Transaction simulatedExtractTransaction = TransactionManager.open(transaction)) {
+                    extracted = from.extract(index, fromResource, amount - totalMoved, simulatedExtractTransaction);
+                }
+
+                try (Transaction transferTransaction = TransactionManager.open(transaction)) {
+                    // check how much can be inserted
+                    int inserted = to.insert(fromResource, extracted, transferTransaction);
+
+                    // extract it, or rollback if the amounts don't match
+                    if (from.extract(index, fromResource, inserted, transferTransaction) == inserted) {
+                        totalMoved += inserted;
+                        transferTransaction.commit();
+                        lastMovedResource = fromResource;
+                    }
+                }
+
+                if (amount == totalMoved) {
+                    // early return if nothing can be moved anymore
+                    return stackFactory.create(lastMovedResource, totalMoved);
+                }
+            }
+            return stackFactory.create(totalMoved == 0 ? defaultResource : lastMovedResource, totalMoved);
+        } catch (Exception e) {
+            CrashReport report = CrashReport.forThrowable(e, "Moving resources between storages");
+            //noinspection DataFlowIssue
+            report.addCategory("Move details")
+                    .setDetail("Input", from::toString)
+                    .setDetail("Output", to::toString)
+                    .setDetail("Filter", filter::toString)
+                    .setDetail("Amount", amount)
+                    .setDetail("Transaction", transaction);
+            throw new ReportedException(report);
+        }
+    }
+
+    public static <T extends IResource> long getAmountAsLong(IResourceHandler<T> handler) {
+        long sum = 0L;
+        int size = handler.size();
+        for (int index = 0; index < size; index++) {
+            sum += handler.getAmountAsLong(index);
+            if (sum < 0) return Long.MAX_VALUE;
+        }
+        return sum;
+    }
+
+    public static <T extends IResource> long getCapacityAsLong(IResourceHandler<T> handler) {
```

Comment:
Should we provide an overload of this where they can specify the resource? Such as if it is a smaller amount once we have an item stored, and then they want to know how much of the item they would be able to insert?